### PR TITLE
Add v3 and v4 encodings, using mochilo

### PR DESCRIPTION
**Pull request by spraints**
This isn't quite done yet, but I wanted to open up this PR to capture the progress so far. This adds a new BERT encoding (v3) that's based on msgpack. This is much faster, especially for long arrays.

```
                                     user     system      total        real
BERT v1 tiny                     0.020000   0.000000   0.020000 (  0.019957)
BERT v1 small                    0.150000   0.000000   0.150000 (  0.141066)
BERT v1 large                    0.710000   0.000000   0.710000 (  0.723820)
BERT v1 complex                  2.950000   0.000000   2.950000 (  2.972712)
BERT v1 long array               3.870000   0.000000   3.870000 (  3.898899)
BERT v2 tiny                     0.010000   0.000000   0.010000 (  0.017890)
BERT v2 small                    0.160000   0.000000   0.160000 (  0.159581)
BERT v2 large                    0.720000   0.000000   0.720000 (  0.733793)
BERT v2 complex                  2.960000   0.010000   2.970000 (  2.975499)
BERT v2 long array               4.080000   0.000000   4.080000 (  4.106795)
BERT v3 tiny                     0.020000   0.000000   0.020000 (  0.024472)
BERT v3 small                    0.020000   0.000000   0.020000 (  0.025297)
BERT v3 large                    0.320000   0.020000   0.340000 (  0.337157)
BERT v3 complex                  0.090000   0.000000   0.090000 (  0.085884)
BERT v3 long array               0.100000   0.000000   0.100000 (  0.104439)
```

String encodings are currently not working. I've heard that bananapack (https://github.com/brianmario/mochilo) could solve this. I skimmed through the mochilo code, and I'm not sure that it'll work, because it lacks the [custom type stuff](https://github.com/spraints/bert/blob/5c88d933e09df9314ca1a0b6bffdd70edf6d629c/lib/bert/msgpack.rb).

cc @carlosmn @piki @vmg

### CODE REVIEW

File: bert.gemspec
Comment (piki) [inline bert.gemspec]: Remove `byebug` before merging, I assume?

**Comment by carlosmn** [comment]
> I skimmed through the mochilo code, and I'm not sure that it'll work, because it lacks the custom type stuff.

In that case, it sounds like we'll have to build it ourselves.

Since we already have `Encoder.convert` to convert the ruby objects into whatever we need to pass though, we should be able to return some `EncodedString` type and register our own type, which packs as a tuple of contents plus encoding. That should be enough to work around msgpack always trying to use its own type for strings, no?

**Comment by carlosmn** [comment]
That seems to work out well enough. It's presumably worth not returning `EncodedString` if we see it's binary or UTF-8 or whatever msgpack can do itself.

```diff
diff --git a/lib/bert.rb b/lib/bert.rb
index 352016a..be499f7 100644
--- a/lib/bert.rb
+++ b/lib/bert.rb
@@ -30,3 +30,17 @@ require 'bert/decoder'
 def t
   BERT::Tuple
 end
+
+class EncodedString
+  def initialize(string)
+    @string = string
+  end
+
+  def raw
+    @string.byteslice(0..-1)
+  end
+
+  def encoding
+    @string.encoding
+  end
+end
diff --git a/lib/bert/encoder.rb b/lib/bert/encoder.rb
index 3e9bd0b..6e4bb77 100644
--- a/lib/bert/encoder.rb
+++ b/lib/bert/encoder.rb
@@ -19,7 +19,15 @@ module BERT
     #
     # Returns the converted Ruby object
     def self.convert(item)
-      return item if Encode.version == :v3
+      if Encode.version == :v3
+        case item
+        when String
+          return EncodedString.new(item)
+        else
+          return item
+        end
+      end
+
       case item
         when Hash
           pairs = []
diff --git a/lib/bert/msgpack.rb b/lib/bert/msgpack.rb
index 49ff96e..600844d 100644
--- a/lib/bert/msgpack.rb
+++ b/lib/bert/msgpack.rb
@@ -11,6 +11,9 @@ module BERT
     factory.register_type 0x03, Time,
       :packer => lambda { |time| [time.to_f].pack("G") },
       :unpacker => lambda { |raw| Time.at(raw.unpack("G").first) }
+    factory.register_type 0x04, EncodedString,
+      :packer => lambda { |s| [s.encoding.name, s.raw].pack("Z*a*") },
+      :unpacker => lambda { |raw| enc, str = raw.unpack("Z*a*"); str.force_encoding(enc); str }
     return factory
   end
 end
```

**Comment by spraints** [comment]
Does the `EncodedString` thing work for strings that are nested in an array or hash?

I tried two other implementations: one using a modified version of msgpack that allows customizations of the String serializer (https://github.com/msgpack/msgpack-ruby/compare/msgpack:808fdeb...spraints:8a2f5bf), and one using a modified version of mochilo that allows custom types to be registered (https://github.com/brianmario/mochilo/compare/brianmario:ced75e5...spraints:458fcb5). I like the mochilo version better because it's faster, and I think it'll be easier to get a PR merged.

For now, the msgpack version is called "v3" and the mochilo version is called "v4", so that I can compare them side-by-side. Whichever one wins will end up being called "v3".

**Comment by carlosmn** [comment]
>Does the EncodedString thing work for strings that are nested in an array or hash?

Good point. The code as posted won't do that (I guess we should add a test ;), we'd have to do something similar to the handling in actual BERT where we `item.map { |i| convert(i) }`.

If we can get mochilo to accept custom types, I do think that sounds better, as it saves us from handling the encoding in this gem and would let us re-use that wherever we want mgpack in the future.

**Comment by spraints** [comment]
This is now using brianmario/mochilo#19, which handles the extra types that BERT needs in C, and v3 is now even faster. (Note that I changed the benchmark a little bit, because the 45s test cases were annoying to wait for.) Here are the times for the BERT's Ruby decoders:

```
                                     user     system      total        real
BERT v1 tiny                     0.010000   0.000000   0.010000 (  0.013944)
BERT v1 small                    0.130000   0.000000   0.130000 (  0.124755)
BERT v1 large                    0.560000   0.010000   0.570000 (  0.571472)
BERT v1 complex                  2.750000   0.000000   2.750000 (  2.758103)
BERT v1 long array               5.440000   0.000000   5.440000 (  5.476571)
BERT v2 tiny                     0.020000   0.000000   0.020000 (  0.017456)
BERT v2 small                    0.140000   0.000000   0.140000 (  0.136073)
BERT v2 large                    0.640000   0.000000   0.640000 (  0.648367)
BERT v2 complex                  2.610000   0.000000   2.610000 (  2.627479)
BERT v2 long array               5.490000   0.010000   5.500000 (  5.519414)
BERT v3 tiny                     0.010000   0.000000   0.010000 (  0.003783)
BERT v3 small                    0.000000   0.000000   0.000000 (  0.005775)
BERT v3 large                    0.240000   0.010000   0.250000 (  0.251264)
BERT v3 complex                  0.040000   0.000000   0.040000 (  0.037354)
BERT v3 long array               0.180000   0.000000   0.180000 (  0.183192)
```

Here are the times for BERT's C decoders:

```
                                     user     system      total        real
BERT v1 tiny                     0.000000   0.000000   0.000000 (  0.009276)
BERT v1 small                    0.070000   0.000000   0.070000 (  0.091471)
BERT v1 large                    0.310000   0.030000   0.340000 (  0.342087)
BERT v1 complex                  1.130000   0.010000   1.140000 (  1.144703)
BERT v1 long array               2.240000   0.010000   2.250000 (  2.257003)
BERT v2 tiny                     0.010000   0.000000   0.010000 (  0.008312)
BERT v2 small                    0.060000   0.000000   0.060000 (  0.060999)
BERT v2 large                    0.320000   0.020000   0.340000 (  0.346957)
BERT v2 complex                  1.140000   0.000000   1.140000 (  1.135720)
BERT v2 long array               2.020000   0.000000   2.020000 (  2.037739)
BERT v3 tiny                     0.000000   0.000000   0.000000 (  0.003123)
BERT v3 small                    0.000000   0.000000   0.000000 (  0.003621)
BERT v3 large                    0.240000   0.040000   0.280000 (  0.289285)
BERT v3 complex                  0.040000   0.000000   0.040000 (  0.036944)
BERT v3 long array               0.170000   0.000000   0.170000 (  0.170969)
```

**Comment by spraints** [comment]
This is ready for review.

**Review by carlosmn — APPROVED** [review]
Makes sense to me. The only think I'm wondering about is how we'd end up shipping the mochilo gem. We'd need to vendor it in the app, no?

**Comment by spraints** [comment]
In the app yes, I'll bundle it. Here I was planning to leave it bundled via git, but I could put the gem files in vendor/cache. The *right* answer is probably to publish the gem updates.

**Comment by piki** [comment]
Looks fine to me.  I'm curious, why are there both C and pure-Ruby implementations of the decoder?

**Review by piki — APPROVED** [review]

**Comment by spraints** [comment]
> I'm curious, why are there both C and pure-Ruby implementations of the decoder?

I don't know. That goes back to a commit 8y ago that says "remove reliance on [erlectricity](https://github.com/mojombo/erlectricity)" (41503465479e8762916d6997d91639f0d7308a13). 

> Only those with write access to this repository can merge pull requests.

cc @charliesome @brianmario @carlosmn This PR is ready to be merged, but I don't have a green merge button.

**Pull request merged**
